File: 0167-two-sum-ii-input-array-is-sorted/0167-two-sum-ii-input-array-is-sorted.js
/**
 * @param {number[]} numbers
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(numbers, target) {
    /*
    - input arr is sorted already in increasing order
    - there is an one solution exactly
    - cant use the same element twice
    - constant extra space
    
    ex) 
           0 1 2
    arr = [2,3,4] t = 6
    => [1,3]
    
    */
    
    let left = 0;
    let right = numbers.length-1;
    while(left < right){
        if(numbers[left] + numbers[right] === target) return [left+1, right+1];
        else if(numbers[left] + numbers[right] > target) right--;
        else left++
    }
};

File: 0167-two-sum-ii-input-array-is-sorted/0167-two-sum-ii-input-array-is-sorted.py
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        '''
        U:
        q) can arr be empty?
        q) can arr have negative nums?
        ex) [2,7,11,15] t=9
            
        M: two ptrs
        
        P: 
            1. init two ptrs (left and right)
            2. while left< right:
                2.1) if arr[left] + arr[right] < target, then left++
                2.2) elif arr[left] + arr[right] > target, then right--
                2.3) else return [left+1, right+1]
        '''
        
        left, right = 0, len(numbers)-1
        
        while left < right:
            curSum = numbers[left] + numbers[right]
            if curSum < target: left+=1
            elif curSum > target: right-=1
            else: return [left+1, right+1]